test_plans/weak_object.asciidoc

```diff
@@ -0,0 +1,118 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for weak object
+
+This is a test plan for the APIs described in
+https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_weak_object.asciidoc[SYCL_EXT_ONEAPI_WEAK_OBJECT.asciidoc]
+
+== Testing scope
+
+=== Device coverage
+
+All of the tests described below are performed only on the default device that
+is selected on the CTS command line.
+
+=== Feature test macro
+
+All of the tests should use `#ifdef SYCL_EXT_ONEAPI_WEAK_OBJECT` so they can be skipped
+if feature is not supported.
+
+=== Type coverage
+All of the tests for `weak_object` described below are performed using each of the following `typename SyclObject`:
+
+* `accessor<int>` +
+* `buffer<int>` +
+* `context` +
+* `device` +
+* `event` +
+* `host_accessor<int>` +
+* `kernel` +
+* `kernel_id` +
+* `kernel_bundle<bundle_state::executable>` +
+* `local_accessor<int>` +
+* `platform` +
+* `queue` +
+* `stream` +
+
+== Objects creation
+
+All of the tests for `weak_object` described below using empty or non-empty objects:
+
+* Empty `weak_object` created by using `constexpr weak_object()` constructor
+
+* Non-empty `weak_object` created by using `weak_object(const SyclObject &SYCLObj)` constructor,
+where `SyclObject` was created using one of the following:
+    ** Default class constructor
+    ** Sample data constructor
+    ** `get_<typenmane>()` member functions from other objects. For example `device = queue.get_device()`
+
+== Tests
+
+=== Constructors
+
+Create instances of `weak_object` using: +
+`constexpr weak_object()` +
+`weak_object(const SyclObject &SYCLObj)` +
+`weak_object(const weak_object &Other)` +
+`weak_object(weak_object &&Other)` +
+
+* Check that `object_type` is the same type with `SyclObject` using `std::is_same_v`
+* Check that `try_lock()` returns same `SyclObject` with same value and type using `==` and `std::is_same_v` respectively for non-empty and `std::nullopt` for empty objects
+* Check that `expired()` returns `true` for empty and `false` for non-empty objects
```

Comment:
Since you are testing `try_lock` here, it seems like you should test `lock` also:

> * If the `weak_object` was constructed from an underlying SYCL object `SyclObject`:
>    * Call `lock` and verify that the value it returns compares equal to `SyclObject`.

It seems like you may as well test `reset` here too:

> * Call `reset`, and then verify that `expired()` returns `false`.

Reply:
Added `lock` testing also with exception catching